Improve documentation for the "Matrix groups" section

ThomasBreuer opened the pull request:
- add introductory text and examples for matrix groups (analogous to the recent changes in the documentation for f.p. groups)
- improve the documentation for sesquilinear forms
- change/fix docstrings, add examples to docstrings
- export `invariant_bilinear_form`, `invariant_quadratic_form`, `invariant_sesquilinear_form` (alternatively, we could remove their documentation from the manual)

ThomasBreuer committed:
Improve documentation for the "Matrix groups" section

- add introductory text and examples for matrix groups
  (analogous to the recent changes in the documentation for f.p. groups)
- improve the documentation for sesquilinear forms
- change/fix docstrings, add examples to docstrings
- export `invariant_bilinear_form`, `invariant_quadratic_form`,
  `invariant_sesquilinear_form`
  (alternatively, we could remove their documentation from the manual)

ThomasBreuer committed:
fix a doctest

fingolfin commented on src/Groups/matrices/form_group.jl:
```diff
@@ -184,7 +184,7 @@ An exception is thrown if `base_ring(G)` is not a finite field with even degree
 
 !!! warning "Note:"
-    At the moment, elements of the generating set are returned of type `mat_elem_type(G)`.
+    At the moment, elements of the generating set are returned of type `mat_elem_type(typeof(G))`.
```
For other similar `*_type` function we actually make sure that one can also pass in instances, for convenience. Maybe a good idea to do that here, too? E.g. we do this in AA:
```julia
elem_type(x) = elem_type(typeof(x))
elem_type(T::DataType) = throw(MethodError(elem_type, (T,)))
```

ThomasBreuer replied:
I found `mat_elem_type` a bit irritating: It is not documented at all, but it appears essentially in these warnings in the documentation.

Now I think that the function is useful. Thus I will document `mat_elem_type` and `ring_elem_type`, and then it makes sense to admit also the group as an argument instead of its type.

ThomasBreuer committed:
address comments

- document `ring_elem_type`, `mat_elem_type`, admit a matrix group as argument
- add examples for `corresponding_bilinear_form`, `corresponding_quadratic_form`
- changed `show` for forms to avoid trailing whitespace

ThomasBreuer committed:
change docstrings for examples with random effects

The matrices returned by `corresponding_bilinear_form` and
`corresponding_quadratic_form` are determined at most up to scalars,
thus `jldoctest` cannot be used.

ThomasBreuer commented:
Some tests fail: Aqua's "persistent tasks" complains.
Can the changes from this pull request cause this problem?

fingolfin commented:
I don't think the Aqua issue are from this PR.

This PR is still marked as draft, please let me know when you think it is ready for review.

fingolfin commented on src/Groups/matrices/MatGrp.jl:
````diff
@@ -75,8 +75,42 @@ MatrixGroupElem(G::MatrixGroup{RE,T}, x::T, x_gap::GapObj) where {RE,T} = Matrix
+```jldoctest
+julia> g = GL(2, 3);
+
+julia> ring_elem_type(typeof(g)) == elem_type(typeof(base_ring(g)))
````
So this is also the same as this then?
```suggestion
julia> ring_elem_type(typeof(g)) == elem_type(base_ring_type(g))
```
Makes me wonder if we really need `ring_elem_type`. Ah well, I guess we have it for now...

ThomasBreuer replied:
The change from `typeof(base_ring(X))` to `base_ring_type(X)` would make sense across Oscar (perhaps for another pull request).

fingolfin commented on src/Groups/matrices/forms.jl:
````diff
@@ -194,7 +261,14 @@ end
+defined by `B(u,v) = Q(u+v)-Q(u)-Q(v)`.
+
+# Examples
+```@repl
+Q = quadratic_form(invariant_quadratic_form(GO(3,3)))
+corresponding_bilinear_form(Q)
+```
````
Any reason not to use a doctest here?
````suggestion
```jldoctest
julia> Q = quadratic_form(invariant_quadratic_form(GO(3,3)))
quadratic form with Gram matrix
[0   1   0]
[0   0   0]
[0   0   1]

julia> corresponding_bilinear_form(Q)
symmetric form with Gram matrix
[0   1   0]
[1   0   0]
[0   0   2]
```
````

fingolfin commented on src/Groups/matrices/forms.jl:
````diff
@@ -205,10 +279,17 @@ function corresponding_bilinear_form(B::SesquilinearForm)
 It is defined only in odd characteristic.
+
+# Examples
+```@repl
````
ditto

fingolfin reviewed (approved):
Overall this looks good to me (I would consider changing the examples using `@repl` to use `jldoctest` instead but that's a minor point).

If there is something blocking this, it would be good to know what it is, @ThomasBreuer. Otherwise I hope we can just merge this soonish?

ThomasBreuer committed:
Merge branch 'master' into TB_doc_matgroups_intro
